include/counter.h

```diff
@@ -29,6 +29,88 @@
+ */
+#define COUNTER_ALARM_FLAG_ABSOLUTE BIT(0)
+
+/**
+ * @brief Flag used internally indicating that detection of too late alarm
+ *	  setting is enabled.
+ *
+ * Alarm can be set too late when context is interrupted and counter progresses
+ * beyond requested value.
+ *
+ * Detection of too late setting is vital since if it is not detected alarm
+ * is delayed by full period of the counter (up to 32 bits). Because of
+ * wrapping, it is impossible to distinguish alarm which is short in the past
+ * from alarm which is targeted to expire after full counter period. In order to
+ * detect too late setting, longest possible alarm is limited. When driver can
+ * assume that user will not request alarm which is too far in the future, it
+ * can detect too late setting.
+ *
+ * Example 1:
+ * - Counter top value: 1024
+ * - Current counter value: 20
+ * - Maximal possible alarm value (relative to now): 1000.
+ * - Requested absolute alarm: 23
+ *
+ * Context is interrupted when alarm is being set. Value 23 was set in
+ * peripheral register when counter was at 25. Driver reads counter after
+ * setting the alarm and calculates the difference. Alarm is suppose to expire
+ * 1022 ticks from now (23 - 25 & 1024) which exceeds 1000 limitation.
+ * Late setting is detected.
+ *
+ * Example 2:
+ * - Counter top value: 1024
+ * - Current counter value: 20
+ * - Maximal possible alarm value (relative to now): 1000.
+ * - Requested absolute alarm: 1020
+ *
+ * Context is not interrupted. Alarm is set to 1020. Counter value is 20. Alarm
+ * set correctly. If counter would increase to 21 then difference would be 999
+ * which is below maximal possible alarm value which is considered correct
+ * setting.
+ *
+ * Maximal possible alarm value (relative to now) is related to the longest
+ * potential preemption which can occur when setting the alarm. It may vary
+ * between instances of the driver and is application specific. If counter
+ * period is long (>minutes), it is recommended to set with reserve, e.g. if
+ * counter period is 512 seconds, maximal relative alarm value can be set to
+ * 511 seconds.
+ */
+#define Z_COUNTER_ALARM_FLAG_DETECT_LATE BIT(1)
+
+/**
+ * @brief Counter alarm flag enabling immediate alarm expiration when driver
+ *	  detects that alarm was set too late.
+ *
+ * It is required to distinguish between short alarm which was set too late
+ * and alarm which was long, close to counter range. In order to detect that,
+ * absolute value that can be set is limited to:
+ * @ref counter_read() + @ref counter_get_max_relative_alarm()
+ */
+#define COUNTER_ALARM_FLAG_EXPIRE_WHEN_LATE Z_COUNTER_ALARM_FLAG_DETECT_LATE
+
+/**
+ * @brief Counter alarm flag enabling reporting error when driver detects that
+ *	  alarm was set too late.
+ *
+ * If flag is set then @ref counter_set_channel_alarm returns -ETIME if
+ * requested alarm was set too late.
+ *
+ */
+#define COUNTER_ALARM_FLAG_ERROR_WHEN_LATE \
+	(Z_COUNTER_ALARM_FLAG_DETECT_LATE | BIT(2))
```

Comment:
With such construction of this value, how are the drivers supposed to check if it was selected by the application?
`if ((alarm_cfg->flags & COUNTER_ALARM_FLAG_ERROR_WHEN_LATE) == COUNTER_ALARM_FLAG_ERROR_WHEN_LATE)`?
I think the previous approach was better - one flag that specifies if the detection is to be performed or not, and another flag (valid only together with the first one) if the late-to-set alarm should be signaled with the dedicated error return value.

Reply:
i was thinking that driver can use `Z_COUNTER_ALARM_FLAG_DETECT_LATE` to determine if late setting is used and then in the same condition it can additionally check `COUNTER_ALARM_FLAG_ERROR_WHEN_LATE` to determine if error should be returned.

Reply:
Yes, the first step is obvious: `if (alarm_cfg->flags & Z_COUNTER_ALARM_FLAG_DETECT_LATE)`
but I was asking about this additional check. With the current definition of `COUNTER_ALARM_FLAG_ERROR_WHEN_LATE`  you can't use just `==` (because of other flags that may be added in the future) or just `&` (for this you'd need to use only `BIT(2)`), and the combination of `&` and `==` that I mentioned in my previous comment won't be very convenient nor effective. Is there some other option that I missed?

Reply:
reverted to use two separate flags.